text/0000-cargo-cfg-dependencies.md

```diff
@@ -0,0 +1,168 @@
+- Feature Name: N/A
+- Start Date: 2015-11-10
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Improve the target-specific dependency experience in Cargo by leveraging the
+same `#[cfg]` syntax that Rust has.
+
+# Motivation
+[motivation]: #motivation
+
+Currently in Cargo it's [relatively painful][issue] to list target-specific
+dependencies. This can only be done by listing out the entire target string as
+opposed to using the more-convenient `#[cfg]` annotations that Rust source code
+has access to. Consequently a Windows-specific dependency ends up having to be
+defined for four triples: `{i686,x86_64}-pc-windows-{gnu,msvc}`, and this is
+unfortunately not forwards compatible as well!
+
+[issue]: https://github.com/rust-lang/cargo/issues/1007
+
+As a result most crates end up unconditionally depending on target-specific
+dependencies and rely on the crates themselves to have the relevant `#[cfg]` to
+only be compiled for the right platforms. This experience leads to excessive
+downloads, excessive compilations, and overall "unclean methods" to have a
+platform specific dependency.
+
+This RFC proposes leveraging the same familiar syntax used in Rust itself to
+define these dependencies.
+
+# Detailed design
+[design]: #detailed-design
+
+The target-specific dependency syntax in Cargo will be expanded to to include
+not only full target strings but also `#[cfg]` expressions:
+
+```toml
+[target."cfg(windows)".dependencies]
+winapi = "0.2"
+
+[target."cfg(unix)".dependencies]
+unix-socket = "0.4"
+
+[target."cfg(target_os = \"macos\")".dependencies]
+core-foundation = "0.2"
+```
+
+Specifically, the "target" listed here is considered special if it starts with
+the string "cfg(" and ends with ")". If this is not true then Cargo will
+continue to treat it as an opaque string and pass it to the compiler via
+`--target` (Cargo's current behavior).
+
+> **Note**: There's an [issue open against TOML][toml-issue] to support
+> single-quoted keys allowing more ergonomic syntax in some cases like:
+>
+> ```toml
+> [target.'cfg(target_os = "macos")'.dependencies]
+> core-foundation = "0.2"
+> ```
+
+[toml-issue]: https://github.com/toml-lang/toml/issues/354
+
+Cargo will implement its own parser of this syntax inside the `cfg` expression,
+it will not rely on the compiler itself. The grammar, however, will be the same
+as the compiler for now:
+
+```
+cfg := "cfg(" meta-item * ")"
+meta-item := ident |
+             ident "=" string |
+             ident "(" meta-item * ")"
+```
+
+Like Rust, Cargo will implement the `any`, `all`, and `not` operators for the
+`ident(list)` syntax. The last missing piece is simply understand what `ident`
+and `ident = "string"` values are defined for a particular target. To learn this
+information Cargo will query the compiler via a new command line flag:
+
+```
+$ rustc --print cfg
```

Comment:
Oh wow that's an excellent point! I hadn't even considered that part. This would have the nice benefits of:

* Doesn't require changes to the compiler
* Allows newer Cargo releases to work with older compilers

This does, however, have a drawback that Cargo has to have a whitelist of all possible `#[cfg]` annotations that can possibly be defined. This means that if the compiler adds new ones over time (e.g. all the simd-related ones) or simply adds new values (e.g. a new OS or a new `target_pointer_width`, god forbid), it wouldn't automatically be picked up by Cargo.

In terms of forward compatibility I think I'd prefer to stick with `--print cfg` to be sure that Cargo stays as close to the compiler for as long as possible (although it is likely to inevitably diverge). What do you think though?